File: cxx/tests/src/AllTypesTest.cpp
#include "xplpc/custom/AllTypes.hpp"
#include "fixtures/GeneralTest.hpp"
#include "fixtures/VerifiedCall.hpp"
#include "xplpc/xplpc.hpp"
#include "gtest/gtest.h"

#include <chrono>
#include <limits>
#include <optional>
#include <thread>
#include <vector>

using namespace xplpc::core;
using namespace xplpc::client;
using namespace xplpc::message;

AllTypes createItem()
{
    // The edges are what a serializer loses, since a small value round trips even when the format cannot carry the type.
    return AllTypes{
        std::numeric_limits<int8_t>::min(),
        std::numeric_limits<int16_t>::min(),
        std::numeric_limits<int32_t>::min(),
        9007199254740993,
        0.1f,
        0.1,
        true,
        std::nullopt,
        {},
        {},
        std::chrono::system_clock::time_point{std::chrono::milliseconds(494938800)},
        'z',
        0x00AE,
        0x00AE,
        0x00AE,
        "ok",
    };
}

void validateAllTypes(const AllTypes &allTypes)
{
    auto mapValue = allTypes.typeMap.at("item1");
    auto timestamp = std::chrono::time_point_cast<std::chrono::milliseconds>(allTypes.typeDateTime).time_since_epoch().count();

    EXPECT_EQ(std::numeric_limits<int8_t>::min(), allTypes.typeInt8);
    EXPECT_EQ(std::numeric_limits<int16_t>::min(), allTypes.typeInt16);
    EXPECT_EQ(std::numeric_limits<int32_t>::min(), allTypes.typeInt32);
    EXPECT_EQ(9007199254740993, allTypes.typeInt64);
    EXPECT_EQ(0.1f, allTypes.typeFloat32);
    EXPECT_EQ(0.1, allTypes.typeFloat64);
    EXPECT_EQ(true, allTypes.typeBool);
    EXPECT_EQ(std::nullopt, allTypes.typeOptional);
    EXPECT_EQ(1, allTypes.typeList.size());
    EXPECT_EQ("ok", mapValue);
    EXPECT_EQ(494938800, timestamp);
    EXPECT_EQ('z', allTypes.typeChar);
    EXPECT_EQ(0x00AE, allTypes.typeChar16);
    EXPECT_EQ(0x00AE, allTypes.typeChar32);
    EXPECT_EQ(0x00AE, allTypes.typeWchar);
    EXPECT_EQ("ok", allTypes.typeString);
}

TEST_F(GeneralTest, AllTypesTestSingle)
{
    auto item = createItem();
    item.typeMap["item1"] = "ok";
    item.typeList.push_back(item);

    auto request = Request{"sample.alltypes.single", Param<AllTypes>{"item", item}};

    // clang-format off
    VerifiedCall::run<AllTypes>(request, [](const auto &response) {
        EXPECT_NE(response, std::nullopt);

        if (response)
        {
            auto allTypes = response.value();
            validateAllTypes(allTypes);
        }
    });
    // clang-format on
}

TEST_F(GeneralTest, AllTypesTestSingleAsync)
{
    auto item = createItem();
    item.typeMap["item1"] = "ok";
    item.typeList.push_back(item);

    auto request = Request{"sample.alltypes.single", Param<AllTypes>{"item", item}};

    // clang-format off
    std::thread([=] {
        VerifiedCall::run<AllTypes>(request, [](const auto &response) {
            EXPECT_NE(response, std::nullopt);

            if (response)
            {
                auto allTypes = response.value();
                validateAllTypes(allTypes);
            }
        });
    }).join();
    // clang-format on
}

TEST_F(GeneralTest, AllTypesMultipleTest)
{
    auto item1 = createItem();
    item1.typeMap["item1"] = "xyz";

    auto item2 = createItem();
    item1.typeMap["item2"] = "xyz";

    auto items = std::vector<AllTypes>{item1, item2};

    auto request = Request{"sample.alltypes.list", Param<std::vector<AllTypes>>{"items", items}};

    // clang-format off
    VerifiedCall::run<std::vector<AllTypes>>(request, [](const auto &response) {
        EXPECT_NE(response, std::nullopt);

        auto total = 0;

        if (response)
        {
            auto list = response.value();
            total = list.size();
        }

        EXPECT_EQ(2, total);
    });
    // clang-format on
}

TEST_F(GeneralTest, AllTypesMultipleTestAsync)
{
    auto item1 = createItem();
    item1.typeMap["item1"] = "xyz";

    auto item2 = createItem();
    item1.typeMap["item2"] = "xyz";

    auto items = std::vector<AllTypes>{item1, item2};

    auto request = Request{"sample.alltypes.list", Param<std::vector<AllTypes>>{"items", items}};

    // clang-format off
    std::thread([=] {
        VerifiedCall::run<std::vector<AllTypes>>(request, [](const auto &response) {
            EXPECT_NE(response, std::nullopt);

            auto total = 0;

            if (response)
            {
                auto list = response.value();
                total = list.size();
            }

            EXPECT_EQ(2, total);
        });
    }).join();
    // clang-format on
}

TEST_F(GeneralTest, AllTypesListDoesNotReadPastAListTheCallerSentShort)
{
    // How many items arrive is decided by the caller, so a mapping must not reach for ones that are not there.

    auto request = Request{"sample.alltypes.list", Param<std::vector<AllTypes>>{"items", {}}};

    // clang-format off
    VerifiedCall::run<std::vector<AllTypes>>(request, [](const auto &response) {
        ASSERT_TRUE(response.has_value());
        EXPECT_TRUE(response.value().empty());
    });
    // clang-format on
}

File: kotlin/lib/library/src/androidTest/java/com/xplpc/library/AllTypesTest.kt
package com.xplpc.library

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.xplpc.client.Client
import com.xplpc.message.Param
import com.xplpc.message.Request
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.junit.runner.RunWith
import java.util.Date

@RunWith(AndroidJUnit4::class)
class AllTypesTest {
    data class AllTypes(
        var typeInt8: Byte,
        var typeInt16: Short,
        var typeInt32: Int,
        var typeInt64: Long,
        var typeFloat32: Float,
        var typeFloat64: Double,
        var typeBool: Boolean,
        var typeOptional: Boolean?,
        var typeList: ArrayList<AllTypes>,
        var typeMap: HashMap<String, String>,
        var typeDateTime: Date,
        var typeChar: Char,
        var typeChar16: Char,
        var typeChar32: Char,
        var typeWchar: Char,
        var typeString: String
    )

    private fun createItem(): AllTypes {
        return AllTypes(
            1,
            2,
            3,
            4,
            1.1f,
            2.2,
            true,
            null,
            ArrayList(),
            HashMap(),
            Date(494_938_800),
            'z',
            '®',
            '®',
            '®',
            "ok",
        )
    }

    @Test
    fun singleItem() {
        val item = createItem()
        item.typeList.add(createItem())
        item.typeMap["item1"] = "ok"

        val request = Request("sample.alltypes.single", Param("item", item))

        Client.call<AllTypes>(request) { response ->
            assertNotNull(response)

            response?.let { allTypes ->
                val mapValue = allTypes.typeMap["item1"]

                assertEquals(1.toByte(), allTypes.typeInt8)
                assertEquals(2.toShort(), allTypes.typeInt16)
                assertEquals(3, allTypes.typeInt32)
                assertEquals(4, allTypes.typeInt64)
                assertEquals(1.1f, allTypes.typeFloat32)
                assertEquals(2.2, allTypes.typeFloat64, 0.0)
                assertEquals(true, allTypes.typeBool)
                assertEquals(null, allTypes.typeOptional)
                assertEquals(1, allTypes.typeList.size)
                assertEquals("ok", mapValue)
                assertEquals(494_938_800, allTypes.typeDateTime.time)
                assertEquals('z', allTypes.typeChar)
                assertEquals('®', allTypes.typeChar16)
                assertEquals('®', allTypes.typeChar32)
                assertEquals('®', allTypes.typeWchar)
                assertEquals("ok", allTypes.typeString)
            }
        }
    }

    @Test
    fun singleItemAsync() {
        val item = createItem()
        item.typeList.add(createItem())
        item.typeMap["item1"] = "ok"

        val request = Request("sample.alltypes.single", Param("item", item))

        runBlocking {
            Client.call<AllTypes>(request) { response ->
                assertNotNull(response)

                response?.let { allTypes ->
                    val mapValue = allTypes.typeMap["item1"]

                    assertEquals(1.toByte(), allTypes.typeInt8)
                    assertEquals(2.toShort(), allTypes.typeInt16)
                    assertEquals(3, allTypes.typeInt32)
                    assertEquals(4, allTypes.typeInt64)
                    assertEquals(1.1f, allTypes.typeFloat32)
                    assertEquals(2.2, allTypes.typeFloat64, 0.0)
                    assertEquals(true, allTypes.typeBool)
                    assertEquals(null, allTypes.typeOptional)
                    assertEquals(1, allTypes.typeList.size)
                    assertEquals("ok", mapValue)
                    assertEquals(494_938_800, allTypes.typeDateTime.time)
                    assertEquals('z', allTypes.typeChar)
                    assertEquals('®', allTypes.typeChar16)
                    assertEquals('®', allTypes.typeChar32)
                    assertEquals('®', allTypes.typeWchar)
                    assertEquals("ok", allTypes.typeString)
                }
            }
        }
    }

    @Test
    fun multipleItems() {
        val item1 = createItem()
        item1.typeList.add(createItem())

        val item2 = createItem()
        item2.typeList.add(createItem())

        val request = Request("sample.alltypes.list", Param("items", listOf(item1, item2)))

        Client.call<List<AllTypes>>(request) { response ->
            assertEquals(2, response?.size)
        }
    }

    @Test
    fun multipleItemsAsync() {
        val item1 = createItem()
        item1.typeList.add(createItem())

        val item2 = createItem()
        item2.typeList.add(createItem())

        val request = Request("sample.alltypes.list", Param("items", listOf(item1, item2)))

        runBlocking {
            Client.call<List<AllTypes>>(request) { response ->
                assertEquals(2, response?.size)
            }
        }
    }
}
